### DISCUSSION TITLE
AUR: Warn user when unsupported cflags are enabled

### DISCUSSION THREAD

**Discussion by ScoreUnder** [rocm-arch/rocm-arch#890] [Ideas]
To save the trouble of building for a while and getting an error at the end (#789), would it be possible to get a warning at the start of the PKGBUILD build step if unsupported options are present in the CFLAGS/CPPFLAGS/(if applicable)CXXFLAGS?

example:

```bash
case "$CFLAGS $CPPFLAGS $CXXFLAGS" in
    -fstack-protector*)
        warning 'The stack protector is not supported in rocm, you may need to adjust C*FLAGS for this build to work'
        ;;
esac
```

**Comment by fabiscafe** [comment]
IMHO That would require a lot of extra work all around. The pkgbuild should be tuned to build in a clean chroot, set up by [devtools](https://gitlab.archlinux.org/archlinux/devtools/-/blob/master/config/makepkg/x86_64.conf). It should absolutely not try to workaround user-specific settings.

That said, I'm not sure about warnings, but I think the pkgbuild is still not the best place for this.

**Comment by ScoreUnder** [comment]
I see what you mean, though I think the majority of users will be using an AUR wrapper or makepkg itself, which do not use clean chroots to build. I don't think building in a chroot is common knowledge among casual AUR users.

**Comment by fabiscafe** [comment]
Sure, still you're expected to know how to build without helper. If you can't, maybe you shouldn't go for Arch just now. Please see: https://wiki.archlinux.org/title/Arch_Linux - [https://wiki.archlinux.org/title/Frequently_asked_questions#Why_would_I_not_want_to_use_Arch?](https://wiki.archlinux.org/title/Frequently_asked_questions#Why_would_I_not_want_to_use_Arch?)

There are a few packages, like tooling - [glibc](https://github.com/archlinux/svntogit-packages/blob/packages/glibc/trunk/PKGBUILD#L74) who exclude build flags, but here again these are or where default flags.

I'm against changing things for users ease to build on the distribution end. This should always go either upstream or set by the user themself.

**Reply by ScoreUnder** [reply]
I resent the implication that those using the makepkg tool [as documented](https://wiki.archlinux.org/title/Arch_User_Repository#Build_the_package) are somehow "not ready for" arch - I have been using arch for 9 years now and am always a little frustrated to hear that people see our community as elitist and insular, but comments like this certainly don't help.


If the position of this project is that all AUR users should read the upstream documentation of all their dependencies before build, that is fair enough, but this feature request is simply to save people hours of wasted compute time when an obvious and trivially detectable misconfiguration is present. The stack protector is an essential security feature in many programs, so it is not a particularly uncommon configuration. Given the extreme resource requirement of the build script (over an hour and often far exceeding the 32GB of RAM I have installed) only to have it fail at the last second with an irrecoverable error, I think some sort of safety mechanism, even one that merely warns the user then continues on, is a huge amount of value for a couple of lines of code.

(It is also worth mentioning that even if the user finds some hack to perform incremental compilation to try to 'save' the mistake, the harm done by the stack protector flag is already there and can't easily be reversed without rebuilding the whole thing anyway)

**Reply by fabiscafe** [reply]
It's less of an elitist standpoint, but more of an unwillingness to accept what Arch is. Many people try to make Arch user friendly and I respect that. Arch however is not that. Arch is upstream driven and it should and most likely will not be to place your change will happen. If you want to change something, change it upstream - except if it's something that should be changed on Arch' side.

It's also not about being "ready for Arch". You never are and will hit and overcome new limits as you go using it.

> The stack protector is an essential security feature in many programs, so it is not a particularly uncommon configuration.

I'm with you on that. I would like to have the ability to check on buildsystems side for "minimal requirements". Not only something like "The system needs to have XXGiB of RAM", "Build time will take about YYMin on this system", but also things like "your compiler/settings are unsupported". But all of this requires changes as well as people who test them and should be upstream, not on a distributors end, reachable for a few people, but available by default - for everyone.

------

Another option I see, would be to talk to Arch as well as AUR helper people and introduce a set of AUR-specific options. There are some packages who are only available in the AUR without any chance to go to [community], for reasons. So it might not be completely out of scope to have an `aur-options` array. To define things like 

    aur_options=(cleanbuild heavyusage denybinarydistribution [...] )
    
    cleanbuild => Always build in a clean chroot with Arch default settings
    heavyusage => show a warning that this package will take a long time to build, uses many resources or wont build on weaker systems at all
    denybinarydistribution => forbid distribution of the resulting binary package
    [...]

So that AUR people can define settings the build environment requires.

**Comment by tpkessler** [comment]
Hey @ScoreUnder! That's an interesting idea. Stack protection is certainly the most commonly used flag that is unsupported by hipcc. But I like my PKGBUILDs to be as simple as possible. This means we assume the setup of a minimal Arch Linux (clean chroot) and won't add extra logic to keep track of different user configurations. To quote paragraph 21 in @fabiscafe's link: [RTFM](https://wiki.archlinux.org/title/Arch_terminology#RTFM). For rocm-arch this is the readme of this repo.

**Reply by ScoreUnder** [reply]
I understand, and found the cause of the error by looking at this repo (also worth mentioning that this was not the project I was attempting to install, but a dependency of it). The main motivation here is that the build script will occupy an extreme amount of resources for an hour or more, before finally discovering at the very end (at link time) that an error had occurred right at the beginning. I would consider my PC somewhat powerful -- 32GB RAM, 1st-gen threadripper, nvme drives -- but it is easily ground to almost a halt at several points during the build process, so I think a proactive approach to detecting this would save a whole lot of wasted time and power.

I understand if you don't want to do this, but I am thoroughly convinced of its utility.
